### PR TITLE
Added rule engine controller

### PR DESCRIPTION
**Pull request by irynamatveieva**
## Pull Request description

Added rule engine controller, service and rest call reply node. Also provided corresponding tests for them. 

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [ ] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [ ] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [ ] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.
- [x] If new yml property was added: make sure a description is added (above or near the property).

### CODE REVIEW

File: application/src/main/java/org/thingsboard/server/controller/RuleEngineController.java [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+                public void onSuccess(@Nullable DeferredResult<ResponseEntity> result) {
+                    long expTime = System.currentTimeMillis() + timeout;
+                    HashMap<String, String> metaData = new HashMap<>();
+                    UUID requestId = UUID.randomUUID();
+                    metaData.put("serviceId", serviceInfoProvider.getServiceId());
+                    metaData.put("requestUUID", requestId.toString());
+                    metaData.put("expirationTime", Long.toString(expTime));
+                    TbMsg msg = TbMsg.newMsg(queueName, TbMsgType.REST_API_REQUEST, entityId, currentUser.getCustomerId(), new TbMsgMetaData(metaData), requestBody);
+                    ruleEngineCallService.processRestAPICallToRuleEngine(currentUser.getTenantId(), requestId, msg, queueName != null,
+                            response -> reply(new LocalRequest(msg, currentUser, result), response));
```
Comment (ShvaykaD) [inline application/src/main/java/org/thingsboard/server/controller/RuleEngineController.java:174]:
```suggestion
                            reply -> reply(new LocalRequestMetaData(msg, currentUser, result), reply));
```

File: application/src/main/java/org/thingsboard/server/controller/RuleEngineController.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+                user.getName(),
+                entityId,
+                null,
+                ActionType.REST_API_RULE_ENGINE_CALL,
+                BaseController.toException(e),
+                request,
+                response != null ? response.getData() : "");
+    }
+
+    private record LocalRequest(TbMsg msg, SecurityUser user, DeferredResult<ResponseEntity> responseWriter) {}
```
Comment (ShvaykaD) [inline application/src/main/java/org/thingsboard/server/controller/RuleEngineController.java:236]:
```suggestion
    private record LocalRequest(TbMsg request, SecurityUser user, DeferredResult<ResponseEntity> responseWriter) {}
```

File: application/src/main/java/org/thingsboard/server/service/ruleengine/RuleEngineCallService.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+import org.thingsboard.server.common.msg.TbMsg;
+import org.thingsboard.server.common.msg.queue.TbCallback;
+import org.thingsboard.server.gen.transport.TransportProtos;
+
+import java.util.UUID;
+import java.util.function.Consumer;
+
+public interface RuleEngineCallService {
+
+    void processRestAPICallToRuleEngine(TenantId tenantId, UUID requestId, TbMsg request, boolean useQueueFromTbMsg, Consumer<TbMsg> response);
```
Comment (ShvaykaD) [inline application/src/main/java/org/thingsboard/server/service/ruleengine/RuleEngineCallService.java:28]:
```suggestion
    void processRestAPICallToRuleEngine(TenantId tenantId, UUID requestId, TbMsg request, boolean useQueueFromTbMsg, Consumer<TbMsg> responseConsumer);
```

File: application/src/main/java/org/thingsboard/server/service/ruleengine/DefaultRuleEngineCallService.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+    private final ConcurrentMap<UUID, Consumer<TbMsg>> requests = new ConcurrentHashMap<>();
+
+    public DefaultRuleEngineCallService(TbClusterService clusterService) {
+        this.clusterService = clusterService;
+    }
+
+    @PostConstruct
+    public void initExecutor() {
+        rpcCallBackExecutor = Executors.newSingleThreadScheduledExecutor(ThingsBoardThreadFactory.forName("re-rpc-callback"));
```
Comment (ShvaykaD) [inline application/src/main/java/org/thingsboard/server/service/ruleengine/DefaultRuleEngineCallService.java:54]:
```suggestion
        rpcCallBackExecutor = Executors.newSingleThreadScheduledExecutor(ThingsBoardThreadFactory.forName("re-rest-callback"));
```

File: application/src/main/java/org/thingsboard/server/service/ruleengine/DefaultRuleEngineCallService.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    @PreDestroy
+    public void shutdownExecutor() {
+        if (rpcCallBackExecutor != null) {
+            rpcCallBackExecutor.shutdownNow();
+        }
+    }
+
+    @Override
+    public void processRestAPICallToRuleEngine(TenantId tenantId, UUID requestId, TbMsg request, boolean useQueueFromTbMsg, Consumer<TbMsg> response) {
```
Comment (ShvaykaD) [inline application/src/main/java/org/thingsboard/server/service/ruleengine/DefaultRuleEngineCallService.java:65]:
```suggestion
    public void processRestApiCallToRuleEngine(TenantId tenantId, UUID requestId, TbMsg request, boolean useQueueFromTbMsg, Consumer<TbMsg> response) {
```

File: common/proto/src/main/proto/queue.proto [hunk 6]
```diff
@@ -1465,6 +1471,7 @@ message ToCoreNotificationMsg {
   ToEdgeSyncRequestMsgProto toEdgeSyncRequest = 11;
   FromEdgeSyncResponseMsgProto fromEdgeSyncResponse = 12;
   ResourceCacheInvalidateMsg resourceCacheInvalidateMsg = 13;
+  RestApiCallResponseMsgProto restApiCallResponseMsg = 14;
```
Comment (ShvaykaD) [inline common/proto/src/main/proto/queue.proto:1474]:
```suggestion
  RestApiCallResponseMsgProto restApiCallResponseMsg = 50;
```

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbSendRestApiCallReplyNodeTest.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        when(ctxMock.getRpcService()).thenReturn(rpcServiceMock);
+        String data = """
+                {
+                "temperature": 23,
+                }
+                """;
+        Map<String, String> metadata = Map.of(
+                "requestUUID", "80b7883b-7ec6-4872-9dd3-b2afd5660fa6",
+                "serviceId", "tb-core-0");
+        TbMsg msg = TbMsg.newMsg(null, TbMsgType.REST_API_REQUEST, DEVICE_ID, new TbMsgMetaData(metadata), data);
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbSendRestApiCallReplyNodeTest.java:74]:
```suggestion
        TbMsg msg = TbMsg.newMsg(TbMsgType.REST_API_REQUEST, DEVICE_ID, new TbMsgMetaData(metadata), data);
```

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbSendRestApiCallReplyNodeTest.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+        node.onMsg(ctxMock, msg);
+
+        UUID requestUUID = UUID.fromString("80b7883b-7ec6-4872-9dd3-b2afd5660fa6");
+        verify(rpcServiceMock).sendRestApiCallReply(eq("tb-core-0"), eq(requestUUID), eq(msg));
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbSendRestApiCallReplyNodeTest.java:79]:
```suggestion
        verify(rpcServiceMock).sendRestApiCallReply("tb-core-0", requestUUID, msg);
```

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbSendRestApiCallReplyNodeTest.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+        verify(ctxMock).tellSuccess(eq(msg));
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbSendRestApiCallReplyNodeTest.java:80]:
```suggestion
        verify(ctxMock).tellSuccess(msg);
```

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbSendRestApiCallReplyNodeTest.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    private TbContext ctxMock;
+    @Mock
+    private RuleEngineRpcService rpcServiceMock;
+
+    @BeforeEach
+    void setUp() throws TbNodeException {
+        node = new TbSendRestApiCallReplyNode();
+        config = new TbSendRestApiCallReplyNodeConfiguration().defaultConfiguration();
+        var configuration = new TbNodeConfiguration(JacksonUtil.valueToTree(config));
+        node.init(ctxMock, configuration);
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbSendRestApiCallReplyNodeTest.java:60]: Please add test for default config.

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbSendRestApiCallReplyNodeTest.java [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+        TbMsg msg = TbMsg.newMsg(null, TbMsgType.REST_API_REQUEST, DEVICE_ID, new TbMsgMetaData(metadata), TbMsg.EMPTY_STRING);
+
+        node.onMsg(ctxMock, msg);
+
+        ArgumentCaptor<Throwable> captor = ArgumentCaptor.forClass(Throwable.class);
+        verify(ctxMock).tellFailure(eq(msg), captor.capture());
+        Throwable throwable = captor.getValue();
+        assertThat(throwable).isInstanceOf(RuntimeException.class).hasMessage("Request body is empty!");
+    }
+}
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbSendRestApiCallReplyNodeTest.java:107]: parameterized test with msg metadata and expected error message. There should be 3 test cases.

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbSendRestApiCallReplyNodeTest.java [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+        node.onMsg(ctxMock, msg);
+
+        UUID requestUUID = UUID.fromString("80b7883b-7ec6-4872-9dd3-b2afd5660fa6");
+        verify(rpcServiceMock).sendRestApiCallReply(eq("tb-core-0"), eq(requestUUID), eq(msg));
+        verify(ctxMock).tellSuccess(eq(msg));
+    }
+
+    @Test
+    void givenRequestIdIsNotPresent_whenOnMsg_thenTellFailure() {
+        TbMsg msg = TbMsg.newMsg(null, TbMsgType.REST_API_REQUEST, DEVICE_ID, TbMsgMetaData.EMPTY, TbMsg.EMPTY_STRING);
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbSendRestApiCallReplyNodeTest.java:85]:
```suggestion
        TbMsg msg = TbMsg.newMsg(TbMsgType.REST_API_REQUEST, DEVICE_ID, TbMsgMetaData.EMPTY, TbMsg.EMPTY_STRING);
```

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbSendRestApiCallReplyNodeTest.java [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        ArgumentCaptor<Throwable> captor = ArgumentCaptor.forClass(Throwable.class);
+        verify(ctxMock).tellFailure(eq(msg), captor.capture());
+        Throwable throwable = captor.getValue();
+        assertThat(throwable).isInstanceOf(RuntimeException.class).hasMessage("Request id is not present in the metadata!");
+    }
+
+    @Test
+    void givenRequestBodyIsEmpty_whenOnMsg_thenTellFailure() {
+        Map<String, String> metadata = Map.of("requestUUID", "80b7883b-7ec6-4872-9dd3-b2afd5660fa6");
+        TbMsg msg = TbMsg.newMsg(null, TbMsgType.REST_API_REQUEST, DEVICE_ID, new TbMsgMetaData(metadata), TbMsg.EMPTY_STRING);
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbSendRestApiCallReplyNodeTest.java:98]:
```suggestion
        TbMsg msg = TbMsg.newMsg(TbMsgType.REST_API_REQUEST, DEVICE_ID, new TbMsgMetaData(metadata), TbMsg.EMPTY_STRING);
```

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbSendRestApiCallReplyNodeTest.java (hunk 10 above)
Comment (ShvaykaD) [reply rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbSendRestApiCallReplyNodeTest.java:60]: also please add a test when the user set non-default fields:

        configuration.setRequestIdMetaDataAttribute("some_custom_request_id_field");
        configuration.setServiceIdMetaDataAttribute("some_custom_service_id_field");

File: application/src/test/java/org/thingsboard/server/controller/RuleEngineControllerTest.java [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+        loginSysAdmin();
+        TbMsg msg = TbMsg.newMsg(null, TbMsgType.REST_API_REQUEST, TenantId.SYS_TENANT_ID, null, TbMsgMetaData.EMPTY, REQUEST_BODY);
+
+        doAnswer(invocation -> {
+            Consumer<TbMsg> consumer = invocation.getArgument(4);
+            consumer.accept(msg);
+            return null;
+        }).when(ruleEngineCallService).processRestAPICallToRuleEngine(eq(TenantId.SYS_TENANT_ID), any(UUID.class), any(TbMsg.class), anyBoolean(), any());
+
+        MvcResult result = doPost("/api/rule-engine/", (Object) REQUEST_BODY).andReturn();
```
Comment (ShvaykaD) [inline application/src/test/java/org/thingsboard/server/controller/RuleEngineControllerTest.java:66]: we need to add new method. doPostAsyncWithTypedResponse. See doPostWithTypedResponse as an example.

Comment (ShvaykaD) [reply application/src/test/java/org/thingsboard/server/controller/RuleEngineControllerTest.java:66]: should be:

```
        ResponseEntity response = doPostAsyncWithTypedResponse("/api/rule-engine/", REQUEST_BODY, new TypeReference<>() {
        }, status().isOk());
```

Comment (ShvaykaD) [reply application/src/test/java/org/thingsboard/server/controller/RuleEngineControllerTest.java:66]: We need to add to the AbstractWebTest logic that will set sys admin user id to some variable. Use @BeforeClass  to do that.

File: application/src/test/java/org/thingsboard/server/controller/RuleEngineControllerTest.java [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+
+        ResponseEntity response = (ResponseEntity) result.getAsyncResult();
+        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
+        assertThat(Objects.requireNonNull(response.getBody()).toString()).isEqualTo(REQUEST_BODY);
+        ArgumentCaptor<TbMsg> captor = ArgumentCaptor.forClass(TbMsg.class);
+        verify(ruleEngineCallService).processRestAPICallToRuleEngine(eq(TenantId.SYS_TENANT_ID), any(), captor.capture(), eq(false), any());
+        TbMsg tbMsg = captor.getValue();
+        assertThat(tbMsg.getData()).isEqualTo(REQUEST_BODY);
+        assertThat(tbMsg.getType()).isEqualTo(msg.getType());
```
Comment (ShvaykaD) [inline application/src/test/java/org/thingsboard/server/controller/RuleEngineControllerTest.java:75]: We need to verify that tbMsg.getOriginator() is sys admin user id

File: application/src/test/java/org/thingsboard/server/controller/RuleEngineControllerTest.java [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+import static org.mockito.ArgumentMatchers.anyBoolean;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@DaoSqlTest
+public class RuleEngineControllerTest extends AbstractControllerTest {
```
Comment (ShvaykaD) [inline application/src/test/java/org/thingsboard/server/controller/RuleEngineControllerTest.java:48]: in each test we need to check that auditLogService is called. We should have some existing methods in parent classes that should check this. See AbstractNotifyEntityTest.

File: application/src/test/java/org/thingsboard/server/controller/RuleEngineControllerTest.java [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+
+        verifyNoMoreInteractions(ruleEngineCallService);
+    }
+
+    @Test
+    public void testHandleRuleEngineRequestWithAuthorityCustomerUser() throws Exception {
+        loginTenantAdmin();
+        Device device = createDevice("test", "123");
+        String deviceIdStr = String.valueOf(device.getId().getId());
+        doPost("/api/customer/" + customerId.getId() + "/device/" + deviceIdStr).andReturn();
```
Comment (ShvaykaD) [inline application/src/test/java/org/thingsboard/server/controller/RuleEngineControllerTest.java:194]: please move to the method in the abstract class at the same level as the createDevice("test", "123"); method.

Comment (ShvaykaD) [reply application/src/test/java/org/thingsboard/server/controller/RuleEngineControllerTest.java:194]: assignDeviceToCustomer

Comment (ShvaykaD) [inline application/src/test/java/org/thingsboard/server/controller/RuleEngineControllerTest.java:194]: do not do andReturn() if it isn't needed.

File: application/src/test/java/org/thingsboard/server/controller/RuleEngineControllerTest.java [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+        loginCustomerUser();
+
+        TbMsg msg = TbMsg.newMsg(null, TbMsgType.REST_API_REQUEST, device.getId(), null, TbMsgMetaData.EMPTY, REQUEST_BODY);
+
+        doAnswer(invocation -> {
```
Comment (ShvaykaD) [inline application/src/test/java/org/thingsboard/server/controller/RuleEngineControllerTest.java:199]: let's move this to the method. We call this in 4 places. Something like mockSuccessfulRestApiCallToRuleEngine.

File: application/src/test/java/org/thingsboard/server/service/queue/DefaultTbCoreConsumerServiceTest.java [hunk 19]
```diff
@@ -529,4 +534,18 @@ public void givenStatsDisabled_whenForwardingInactivityMsgToStateService_thenSta
         then(statsMock).should(never()).log(inactivityMsg);
     }
 
+    @Test
+    public void givenRestApiCallResponseMsgProto_whenForwardToRuleEngineCallService_thenCallOnQueueMsg() {
+        // GIVEN
+        ReflectionTestUtils.setField(defaultTbCoreConsumerServiceMock, "ruleEngineCallService", ruleEngineCallServiceMock);
+        var restApiCallResponseMsgProto = TransportProtos.RestApiCallResponseMsgProto.newBuilder().build();
```
Comment (ShvaykaD) [inline application/src/test/java/org/thingsboard/server/service/queue/DefaultTbCoreConsumerServiceTest.java:541]: .newBuilder().build() -> defaultInstance

File: application/src/test/java/org/thingsboard/server/service/queue/DefaultTbCoreConsumerServiceTest.java [hunk 20]
```diff
@@ -529,4 +534,18 @@ public void givenStatsDisabled_whenForwardingInactivityMsgToStateService_thenSta
...
+        doCallRealMethod().when(defaultTbCoreConsumerServiceMock).forwardToRuleEngineCallService(restApiCallResponseMsgProto, tbCallbackMock);
+        doNothing().when(ruleEngineCallServiceMock).onQueueMsg(restApiCallResponseMsgProto, tbCallbackMock);
```
Comment (ShvaykaD) [inline application/src/test/java/org/thingsboard/server/service/queue/DefaultTbCoreConsumerServiceTest.java:543]: Does it do something?

File: application/src/test/java/org/thingsboard/server/service/rpc/DefaultTbRuleEngineRpcServiceTest.java [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.UUID;
+
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.ArgumentMatchers.isNull;
+import static org.mockito.BDDMockito.then;
+import static org.mockito.Mockito.doCallRealMethod;
+import static org.mockito.Mockito.doNothing;
+
+@ExtendWith(MockitoExtension.class)
+class DefaultTbRuleEngineRpcServiceTest {
```
Comment (ShvaykaD) [inline application/src/test/java/org/thingsboard/server/service/rpc/DefaultTbRuleEngineRpcServiceTest.java:39]: Let's create an instance of TbRuleEngineDeviceRpcService

Comment (ShvaykaD) [reply application/src/test/java/org/thingsboard/server/service/rpc/DefaultTbRuleEngineRpcServiceTest.java:39]: Try  @InjectMocks

File: application/src/test/java/org/thingsboard/server/service/ruleengine/DefaultRuleEngineCallServiceTest.java [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+        UUID requestId = UUID.randomUUID();
+        metaData.put("serviceId", "core");
+        metaData.put("requestUUID", requestId.toString());
+        metaData.put("expirationTime", Long.toString(expTime));
+        TbMsg msg = TbMsg.newMsg("main", TbMsgType.REST_API_REQUEST, TENANT_ID, null, new TbMsgMetaData(metaData), "{\"key\":\"value\"}");
+
+        doNothing().when(tbClusterServiceMock).pushMsgToRuleEngine(any(), any(), any(),anyBoolean(), any());
+        ruleEngineCallService.processRestAPICallToRuleEngine(TENANT_ID, requestId, msg, true, TbMsg::getData);
+
+        assertThat(requests.size()).isEqualTo(1);
```
Comment (ShvaykaD) [inline application/src/test/java/org/thingsboard/server/service/ruleengine/DefaultRuleEngineCallServiceTest.java:103]: verify the value of request

File: application/src/test/java/org/thingsboard/server/service/ruleengine/DefaultRuleEngineCallServiceTest.java [hunk 23]
```diff
@@ -0,0 +1,161 @@
...
+        doAnswer(invocation -> {
+            //check the presence of request in the map after put()
+            assertThat(requests.size()).isEqualTo(1);
+            return null;
+        }).when(tbClusterServiceMock).pushMsgToRuleEngine(any(), any(), any(), anyBoolean(), any());
+        ruleEngineCallService.processRestAPICallToRuleEngine(TENANT_ID, requestId, msg, true, TbMsg::getData);
+
+        verify(tbClusterServiceMock).pushMsgToRuleEngine(eq(TENANT_ID), eq(TENANT_ID), eq(msg), eq(true), isNull());
+        //check map is empty after scheduleTimeout()
+        assertThat(requests.size()).isEqualTo(0);
```
Comment (ShvaykaD) [inline application/src/test/java/org/thingsboard/server/service/ruleengine/DefaultRuleEngineCallServiceTest.java:127]: Awaitility.await("Await until request was deleted from map due to timeout")
                .pollDelay(25, TimeUnit.MILLISECONDS)
                .atMost(30, TimeUnit.SECONDS)
                .until(requests::isEmpty);

**Review by ShvaykaD — CHANGES_REQUESTED** [review]
Please see my comments added above.

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbSendRestApiCallReplyNodeTest.java (hunk 11 above)
Comment (irynamatveieva) [reply rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbSendRestApiCallReplyNodeTest.java]: Need to add 4 test cases here (also including test case with message data)

**Pull request closed without merging**